File: NodeParser/tools/VersionParser.py
import sys
import json

json1 = {
    "Jan": 1,
    "Feb": 2,
    "Mar": 3,
    "Apr": 4,
    "May": 5,
    "Jun": 6,
    "Jul": 7,
    "Aug": 8,
    "Sep": 9,
    "Oct": 10,
    "Nov": 11,
    "Dec": 12
}

resJson = {}

def get_valorant_version(path = "{0}/ShooterGame/Binaries/Win64/VALORANT-Win64-Shipping.exe".format(sys.argv[1])):
    with open(path, "rb") as exe_file:
        # Manifest ID
        file = open('./files/manifest/VALORANT.json')
        manifest = json.load(file)
        resJson["manifestId"] = manifest["Manifest ID"]

        # Parse EXE file
        data = exe_file.read()
        pattern = "++Ares-Core+release-".encode("utf-16le")
        patternBuild = "ProductVersion".encode("utf-16le")

        # Formatting Data
        pos = data.find(pattern) + len(pattern)
        fullData = data[pos:pos+100].decode("utf-16le").rstrip("\x00").split("\u0000")
        parsedData = []
        for i in range(0, len(fullData)):
            if fullData[i] != "":
                parsedData.append(fullData[i])

        # VALORANT Branch
        branch = "release-{0}".format(parsedData[0].rstrip("\x00"))
        resJson["branch"] = branch

        # Build Date
        date = parsedData[1].rstrip("\x00").split(" ")
        parsedDate = []
        for i in range(0, len(date)):
            if date[i] != "":
                parsedDate.append(date[i])
        month =  parsedDate[0]
        day =  parsedDate[1]
        year =  parsedDate[2]
        buildDate = "{0}-{1}-{2}T00:00:00.000Z".format(add_zero(year), add_zero(json1[month]), add_zero(day))

        # VALORANT Version
        if "." in parsedData[2].rstrip("\x00"):
            version = parsedData[2].rstrip("\x00")
        else:
            version = parsedData[3].rstrip("\x00")
        resJson["version"] = version

        # Build Version
        pos = data.find(patternBuild) + len(patternBuild) + 2
        buildVersion = data[pos:pos+4].decode("utf-16le").rstrip("\x00")
        resJson["buildVersion"] = buildVersion

        # Riot Client Version
        riotClientVersion = "{0}-shipping-{1}-{2}".format(branch, buildVersion, version.split(".").pop())
        resJson["riotClientVersion"] = riotClientVersion

        # Riot Client Build
        get_riot_client_version()
        resJson["buildDate"] = buildDate.replace(" ", "0")
        save_json()

def get_riot_client_version(path = "{0}/RiotClientServices.exe".format(sys.argv[2]), path1 = "{0}/RiotGamesApi.dll".format(sys.argv[2])):
    with open(path, "rb") as exe_file:
        # Riot Client Build
        data = exe_file.read()
        pattern = "FileVersion".encode("utf-16le")

        pos = data.find(pattern) + len(pattern)
        fullData = data[pos:pos+100].decode("utf-16le").rstrip("\x00").split("\u0000")
        parsedData = []
        for i in range(0, len(fullData)):
            if fullData[i] != "":
                parsedData.append(fullData[i])

        riotClientVersion = parsedData[0]
        
        # Riot Api Version
        with open(path1, "rb") as exe_file:
            data = exe_file.read()
            pattern = "FileVersion".encode("utf-16le")

            pos = data.find(pattern) + len(pattern)
            fullData = data[pos:pos+100].decode("utf-16le").rstrip("\x00").split("\u0000")
            parsedData = []
            for i in range(0, len(fullData)):
                if fullData[i] != "":
                    parsedData.append(fullData[i])
            apiVersion = parsedData[0]
            apiVersion = apiVersion.split(".")[3]

            riotClientBuild = riotClientVersion + "." + apiVersion
            resJson["riotClientBuild"] = riotClientBuild

def save_json():
    with open("./files/version.json", "w") as outfile:
        json.dump(resJson, outfile)

def add_zero(number):
    if len(str(number)) == 1:
        return "0" + str(number)
    else:
        return str(number)

get_valorant_version()

File: ValoParser/Parsers/VersionParser.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.Json.Nodes;
using ValoParser.Utils;

namespace ValoParser.Parsers
{
    public class VersionParser
    {
        public VersionParser() { }

        Dictionary<string, string> json1 = new Dictionary<string, string>
        {
            { "Jan", "01" },
            { "Feb", "02" },
            { "Mar", "03" },
            { "Apr", "04" },
            { "May", "05" },
            { "Jun", "06" },
            { "Jul", "07" },
            { "Aug", "08" },
            { "Sep", "09" },
            { "Oct", "10" },
            { "Nov", "11" },
            { "Dec", "12" }
        };

        private static JsonObject jsonObject = new JsonObject();

        public void GetVersionContent(string gamePath = "C:\\Riot Games\\VALORANT\\live", string apiPath = "C:\\Riot Games\\Riot Client")
        {

            string exePath = gamePath + "\\ShooterGame\\Binaries\\Win64\\VALORANT-Win64-Shipping.exe";

            // VALORANT-Win64-Shipping.exe

            byte[] fileBytesExe = File.ReadAllBytes(exePath);

            byte[] utf16lePatternExe = Encoding.Unicode.GetBytes("++Ares-Core+release-");

            int indexExe = FindPattern(fileBytesExe, utf16lePatternExe);

            if (indexExe != -1 && indexExe + utf16lePatternExe.Length + 100 <= fileBytesExe.Length)
            {
                string text = GetNextBytesAsString(fileBytesExe, indexExe, utf16lePatternExe.Length, 100);

                if (!string.IsNullOrEmpty(text))
                {
                    string[] data = text.Split("\u0000\u0000\u0000");

                    // Branch
                    jsonObject.Add("branch", string.Format("release-{0}", data[0].TrimEnd('\x00')));

                    // Build Date
                    string[] date = data[1].Replace("\u0000", " ").Split(" ");
                    jsonObject.Add("buildDate", string.Format("{0}-{1}-{2}T00:00:00.000Z", date[2].TrimEnd('\x00'), json1[date[0].TrimEnd('\x00')], date[1].TrimEnd('\x00')));

                    // Build Version
                    jsonObject.Add("buildVersion", string.Format("{0}", date[3].TrimEnd('\x00')));

                    // Build Version
                    jsonObject.Add("version", string.Format("{0}", data[2].TrimEnd('\x00')));

                    // Riot Client Version
                    jsonObject.Add("riotClientVersion", string.Format("release-{0}-shipping-{1}-{2}", data[0].TrimEnd('\x00'), date[3].TrimEnd('\x00'), data[2].TrimEnd('\x00').Split(".").Last()));
                }
            }
            else
            {
                Console.WriteLine("VersionUtil: Pattern not found or insufficient bytes available after the pattern. (VALORANT-Win64-Shipping.exe)");
                return;
            }

            // RiotClientServices.exe

            string riotClientBuild = "";

            byte[] fileBytesCli = File.ReadAllBytes(apiPath + "\\RiotClientServices.exe");

            byte[] utf16lePatternCli = Encoding.Unicode.GetBytes("FileVersion");

            int indexCli = FindPattern(fileBytesCli, utf16lePatternCli);

            if (indexCli != -1 && indexCli + utf16lePatternCli.Length + 100 <= fileBytesCli.Length)
            {
                string text = GetNextBytesAsString(fileBytesCli, indexCli, utf16lePatternCli.Length, 100);

                if (!string.IsNullOrEmpty(text))
                {
                    string[] data = text.Split("\u0000\u0000\u0000");
                    riotClientBuild += data[0].Replace("\u0000", " ").TrimStart(' ').Split(" ")[0];
                }
            }
            else
            {
                Console.WriteLine("VersionUtil: Pattern not found or insufficient bytes available after the pattern. (RiotClientServices.exe)");
                return;
            }

            // RiotGamesApi.dll

            byte[] fileBytesApi = File.ReadAllBytes(apiPath + "\\RiotGamesApi.dll");

            byte[] utf16lePatternApi = Encoding.Unicode.GetBytes("FileVersion");

            int indexApi = FindPattern(fileBytesApi, utf16lePatternApi);

            if (indexApi != -1 && indexApi + utf16lePatternApi.Length + 100 <= fileBytesApi.Length)
            {
                string text = GetNextBytesAsString(fileBytesApi, indexApi, utf16lePatternApi.Length, 100);

                if (!string.IsNullOrEmpty(text))
                {
                    string[] data = text.Split("\u0000\u0000\u0000");

                    riotClientBuild += "." + data[0].Replace("\u0000", " ").TrimStart(' ').Split(" ")[0].Split(".").Last();

                    jsonObject.Add("riotClientBuild", riotClientBuild);
                }
            }
            else
            {
                Console.WriteLine("VersionUtil: Pattern not found or insufficient bytes available after the pattern. (RiotGamesApi.dll)");
                return;
            }

            UassetUtil.exportJson(jsonObject, "data/version.json");
            return;
        }

        private int FindPattern(byte[] source, byte[] pattern)
        {

            for (int i = 0; i <= source.Length - pattern.Length; i++)
            {
                bool found = true;
                for (int j = 0; j < pattern.Length; j++)
                {
                    if (source[i + j] != pattern[j])
                    {
                        found = false;
                        break;
                    }
                }

                if (found)
                {
                    return i;
                }
            }

            return -1;
        }

        private string GetNextBytesAsString(byte[] source, int index, int offset, int length)
        {
            if (index + offset + length <= source.Length)
            {
                byte[] nextBytes = new byte[length];
                Array.Copy(source, index + offset, nextBytes, 0, length);

                string text = Encoding.Unicode.GetString(nextBytes);
                return text;
            }

            return null;
        }
    }
}
